CliFx/Suggestions/Design.md

```diff
@@ -0,0 +1,272 @@
+
+```
+### clifx-suggest-begins-here-{commandName}-{scriptVersion}
+<script goes here>
+### clifx-suggest-ends-here-{commandName)";
+```
+
+A .backup file is always created at the profile location.
+
+## Suggest Mode Workflow
+
+1. `[suggest]` directive detected by `CommandInput`
+2. `CliApplication` checks for --install option.
+   1. `SuggestShellHookInstaller` detects available environments, modifies profiles.
+   2. End.
+3. `CliApplication` retrieves suggestions from `SuggestionService`
+   1. `SuggestionService` builds the raw user input from information supplied to `[suggest]` mode. See ExtractCommandText().
+   2. `SuggestionService` splits the user input into arguments.
+   3. `CommandInput.Parse()` parses the arguments, producing a `CommandInput`
+   4. The parse information in `CommandInput` is used to create suggestions.
+4. `CliApplication` outputs suggestions to standard output.
+
+
+# Design Considerations
+
+### Cursor positioning and argument delimination
+
+The command line is supplied to `CommandInput` as an array of strings. The raw command line is not made available via the .NET libraries, posing the following challenges:
+
+1. How to behave when the cursor in the command line is not always at the end of the string.
+
+```ps
+PS C:\> cmd.exe book a
+                # ^ cursor might be here - auto completion should be for "boo"
+```
+2. How to handle difficult Powershell escaping
+
+Consider the following application:
+```cs
+
+public static void Main()
+{
+    foreach(var arg in Environment.GetCommandLineArgs())
+    {
+      Console.WriteLine(arg);
+    }  
+}
+
+```
+Consider the following powershell behaviour
+```ps
+
+PS C:\> cli a b "a b" 'a b' '"a b"' "`"a b`""
+cli
+a
+b
+a b
+a b
+a b # expected "a b" (quotes inclusive)?
+a b # expected "a b" (quotes inclusive)?
+```
+
+The behaviour gets stranger when the following auto-completion hook is considered:
+
+```ps
+$scriptblock = {
+    param($wordToComplete, $commandAst, $cursorPosition)        
+        $command = "CliFx.Demo"
+
+        $result = &$command `[suggest`] $commandAst     # note: $commandAst passed as single parameter
+        Set-content -path C:\temp\log.txt -value $result 
+}
+
+Register-ArgumentCompleter -Native -CommandName "CliFx.Demo" -ScriptBlock $scriptblock
+```
+
+We get the following output:
+
+```ps
+CliFx.Demo
+[suggest]
+CliFx.Demo a b a
+b 'a b' 'a
+b' `a b`"
+```
+
+Observations:
+1. Text passed to the `[suggest]` directive is split inconsistently when passed via the powershell hook.
+2. Cursor positioning seems difficult to predict in relation to the rest of the supplied terms. Note that CliFx is supplied with command line arguments after processing for whitespace. I couldn't find a platform indepdenant way of extracting the raw command line string. 
```

Comment:
I know on .NET/Powershell `Environment.CommandLine` returns the raw string without processing.